### DISCUSSION TITLE
Refactor playout_controls.sh and other bash scripts

### DISCUSSION THREAD

**Discussion by pabera** [MiczFlor/RPi-Jukebox-RFID#1298] [Ideas]
I wonder if it would be useful to refactor the [playout_controls.sh](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/scripts/playout_controls.sh) file a little bit. With over 1k LOC, it's became pretty large and hard to dive into.

I would volunteer to do it but I could use some heads up whether there are hidden spots to be aware of.

Or is there a special reason that I don't know of why this file hasn't been broken up in smaller junks?

I believe there are other files too that could use some refactoring :-)

Cheers

**Comment by s-martin** [comment]
Hi,
I agree playout_controls.sh is a little hard to work with.

Issue is that this file is sort of the "backbone" of Phoniebox, so if something goes wrong here it probably affects all users.

Nevertheless I agree that some improvements should be done. This file also causes some performance issues.
I would suggest that you try to test all of your improvements thoroughly with your setup as we don't have any  tests here.

Please also have a look at #774 for a more general discussion.

Pinging also @MiczFlor

**Reply by pabera** [reply]
Thanks, I didn't know there was another thread about this topic. This should be closed then.

**Comment by MiczFlor** [comment]
The fact that this script is all one long file comes from the very early days, where I treated this *like* and API. Everything would pipe into that script, the script would do the magic.
Breaking it up might make things easier - or more complicated?
I've been dreaming about an API for a long time, where other apps could use the same point, apps like the web app, the bash scripts, smart home apps, but also custom apps could be written.
Does it make sense to think bigger?

**Reply by pabera** [reply]
> Breaking it up might make things easier - or more complicated?

I like to know more about why you think it could become more complicated? I have obviously not dived too deep into the code yet. Any insights are appreciated.

> Does it make sense to think bigger?

Thinking big is always a great idea.. I am all with you. But constantly thinking big can prevent you from delivering results ;-)

> where other apps could use the same point, apps like the web app, the bash scripts, smart home apps, but also custom apps could be written

An idea could be to abstract those low level functions into an independent library and consume it. But even then, I believe it might a good way to start easy.

**Reply by MiczFlor** [reply]
Ok, I am curious :) go for it, I will help where I can, answering questions. Do you think that's a separate branch?
We'll think bigger when the kids are bigger who started us to indulge in this project in the first place, then the took all the time to keep working on it :)

**Comment by MiczFlor** [comment]
Hi @pabera 
since there is a parallel discussion about the future in another discussion here, I was wondering: did you find time to start the refactoring of the bash script? Anything I could look at and help out with? We should start a new branch in this repo, I think. Possibly you can create a pull request with your work as it is?

**Reply by pabera** [reply]
I have refactored it yes. But I was not able to test it and that is still important to be done. I have two Phonieboxes, but the new one needs to be completely ready before I can wipe the old one :-) Give me a few more days.

I believe this refactor is still useful regardless of the other discussion which I am following closely.
